### ISSUE TITLE
remove jQueryUI Sortable dependency

### ISSUE THREAD

**Issue by kofifus** [6pac/SlickGrid#541]
The doc says dependencies include:

> jQueryUI Sortable (optional, only if column reordering is enabled)

First I find this a bit confusing - jQueryUI Sortable is not a separate library so really the dependency is for jQueryUI. I think the word 'Sortable' should be removed ?

Second and mainly, this is a very unusual dependency to have in 2020 as JqueryUI is an old bug prone library no longer developed (last release 2016) and is known to have various conflicts with other libraries.

Can the Sortable feature be integrated into slickgrid to remove this dependency ?
Thanks

**Comment by 6pac** [comment]
Yes, there is definitely scope for losing a few dependencies - another one is the custom drag/drop libraries. A few have attempted the drag/drop switch, but it turns out that modern drag/drop libs (and native implementations) work very differently and incompletely, so have ended up scrapping the changes.  
It needs someone with a high level of expertise and some time to set aside to make it work. Dropping jQueryUI sortable may be significantly easier (by the way, you can download targeted compiles of jQueryUI with just the modules you need - that is what the docs are referring to). You are welcome to submit a pull request, as I fear that the rest of us are too busy to make this change.

**Comment by kofifus** [comment]
Thanks!

I think there is a bit of a difference here, jquery drag & drop are focused on a specific objective and together are ~9k (minified)  - I really have no problem with that (though why not just include that small code inside slickrun ?)

Whereas jquery-ui.min.js is a general ui library and is ~254k (!).  

I tried the JQuery UI dwonload page to only get 'sortable' (and smoothable theme to make it like the examples), I got a zip with the following files:

images (directory)
jquery-ui.css
jquery-ui.js (now 1KB minified !)
jquery-ui.structure.css
jquery-ui.theme.css

Which of the css are relevant for slickgrid ? 

OK my main question now is can the slickgrid repo in SlickGrid-2.4.30\lib only contain the sortable component as that will reduce the size from 254k to 1k ?

**Comment by 6pac** [comment]
Yep, that is an eminently reasonable request. I think this has slipped through because personally I still use jQueryUI for other controls, eg DatePicker, so I have the full lib anyway.  
Any issue doing this @ghiscoding ?

Re which of the files are directly used by SlickGrid, I'd have to assume all unless I went through the code and narrowed it down. A brute force approach might be removing one file at a time and hard reloading the browser, checking for errors.

Another reason that I haven't worried too much is that I believe browser caching is very good nowadays, so it's likely the file will only be downloaded once.

**Comment by ghiscoding** [comment]
The answer to that question would be... if you find the best optimal and minimal size for jQueryUI and the other dependencies then please share with us and maybe do a Pull Request with what you find to be optimal. However as @6pac mentioned, there are other features of jQueryUI that are used in some of the grids for example the DatePicker is in the `Slick.Editor` but I don't use it in [Angular-Slickgrid](https://github.com/ghiscoding/Angular-Slickgrid), however I do use the jQueryUI AutoComplete and Slider in Angular-Slickgrid, so it all depends what you're going for (are you using `Slick.Editors`) and what lib you're using SlickGrid itself or Angular-Slickgrid/Aurelia-Slickgrid? 

There are other libs, some are even modified for SlickGrid usage, like `jquery.event.drag-2.3.0.js`, `jquery.event.drop-2.3.0.js`, `jquery.mousewheel.js`

**Comment by kofifus** [comment]
thanks @6pac and @ghiscoding 

I think we can differentiate between two kind of features:

There are features where there is a single official way to implement - ie jqureyui for sorting or  jquery.event.drag for dragging. In these cases as these are basic features really I believe it would be best to at least show the minimal code needed as I request for those users who want this feature without the enormous jqueryui baggage.

Then there are features where a user may choose different implementations ie a Datepicker or autocomplete. In these cases I think it is definitely not the case that you want to include the code in ie the lib directory as it belongs in a specific example.

Thanks!



To me sorting is a pretty basic feature of a grid and should be possible without adding a huge UI library that you don't otherwise need. I think that is a reasonable request to at least have an example of !

This is different from fancy things to put inside cells like the date picker you mentioned, here one do want the option of choosing different controllers from different 

Thanks

**Comment by 6pac** [comment]
Agree. There are a number of more modern and lightweight libraries that could do what jQueryUI does. I just want to highlight that the reason we aren't converting to these libraries is mainly because of time to do so, and as a subsidiary reason because of historical baggage.   
If you were to provide a working PR, then we would be most open to the idea.

As a side issue, I agree that jQueryUI has reached its expiry date. However I haven't come across an alternate datepicker that doesn't have a huge library attached to it, and @ghiscoding still uses other controls. These things are evolving all the time, so I'd be interested to hear what you use. If there was another viable option for these other controls, we'd probably be open to that too. Again, it's time to research and test all the possible combinations of controls and libraries.

**Comment by kofifus** [comment]
OK I will close this now. Hopefully I will have time to submit a PR at some point,

**Comment by kofifus** [comment]
@6pac 

> However I haven't come across an alternate datepicker that doesn't have a huge library attached to it

did you see https://github.com/Pikaday/Pikaday ?

**Comment by ghiscoding** [comment]
On one side you're saying that jQuery UI should be removed because it takes too much space but on the other end you suggest to add another external lib. I would think that the first action would be to remove jQueryUI as a dependency but to do that, it would require quite some time which we don't have and that is not our priority. Remember that this is an Open Source project which everyone can contribute, if someone can take the time to drop jQueryUI that would be awesome, Pikaday seems like a nice small lib but we don't want to add another dependency to the lib.

**Comment by kofifus** [comment]
thanks @ghiscoding, agreed. 

I didn't mean pikaday etc should be added to Slickgrid, just showing it is possible to have a date picker which is standalone as 6pac asked.

As far as I can see the _only_ place jquery ui is needed for slickgrid itself is for sortable, so if sortable js+css can be extracted and merged into slickgrid the jquery ui dependcy can be removed. What users chose to put inside cells is up to them and they can use JQuery UI or anything else. However I think all functionality that is of the grid itself (ie sorting columns) should be available without a dependency on a UI lib

**Comment by 6pac** [comment]
Yeah, that datepicker looks good! 
I'd just reiterate that I'm on board with all the ideas mentioned here, it just comes down to a time issue. I think there would be weeks of coding involved to drop jQuery, and days to drop the jQueryUI dependency. However I agree that we should have the most minimal build of jQueryUI included as part of the base build. I'll make that change.

**Comment by ghiscoding** [comment]
> As far as I can see the _only_ place jquery ui is needed for slickgrid itself is for sortable, so if sortable js+css can be extracted and merged into slickgrid the jquery ui dependcy can be removed.

You should re-read what 6pac and I wrote earlier, it's used in much more places than you think. It's also worth to know that when SlickGrid was created, there were no UI lib like Bootstrap or anything there was only jQueryUI and that is why the original author chose it and that is also why it's integrated into the lib and there needs some work to decouple it. 

You should take a look at the source code and perhaps make contributions, I won't contribute on that decoupling since I'm using some of the jQueryUI features in my own lib (since it's a dependency of SlickGrid I might as well use it), for example I use the jQueryUI AutoComplete and Slider and SlickGrid itself uses the DatePicker as well.

**Comment by kofifus** [comment]
thanks @ghiscoding 

I'm surprised, the documentation here https://github.com/6pac/SlickGrid/wiki under "Dependencies" mentions JQueryUI _only_ in relation to sortable.

I have some code that does not need sortable and therefore I did not include JQuery UI so your comment above makes me a bit nervous ... Is this unsafe ?

**Comment by ghiscoding** [comment]
I'm not 100% sure, I never tried to remove it, please go ahead make the try and let us know. If it's only sortable then great, that would be easier to decouple (and you could contribute) if not then perhaps @6pac can update that page.

**Comment by 6pac** [comment]
Just a note ... the ```sortable``` module in jQueryUI is 1200 lines of pretty compact javascript, so I'd tend to think it would be very hard to import/integrate other than just copying it. But in that case, it may as well remain in a minimal jQueryUI library build.

The only real option might be to move to a different library and find something else that will do ``sortable```. But then we go back to the issues of which of probably many competing library to use, and the time to redo the sortable integration.

**Comment by ghiscoding** [comment]
To come back on the subject, I also wanted to see if I could improve my bundle size and in my case I use WebPack bundler with my libs (Angular-Slickgrid, Slickgrid-Universal, ...) and the most optimal way to get smaller size was to `import` only the widgets that I really use.

So I was using `import 'jquery-ui-dist` (from this [npm package](https://www.npmjs.com/package/jquery-ui-dist)) which is a minified version of the **entire** jQueryUI but that does include every widgets (not optimal)... after some trials and error I found the small minimal minimal imports that works with SlickGrid was to import with the following lines (by using [jquery-ui](https://www.npmjs.com/package/jquery-ui) npm package)
```ts
import 'jquery';
import 'jquery-ui/ui/widgets/sortable';
```
But I also use some other widgets for some Editors/Filters and for those specifically, I just import with the Editor/Filter, so for example I have an AutoComplete Editor and on top of that file I just add the autocomplete import
```ts
import 'jquery-ui/ui/widgets/autocomplete';
```
And if you use the Draggable Grouping (or Row Move plugin), then you'll also need these 2 lines
```ts
import 'jquery-ui/ui/widgets/draggable';
import 'jquery-ui/ui/widgets/droppable';
```

I basically followed this Stack Overflow [Answer](https://stackoverflow.com/a/38755491/1212166). 

Does that make my bundle much smaller? It helps a bit, I save about 200-300kb but what I found is that many of these widgets requires other widgets, for example the `controls` widget is installed even though I didn't ask for it, I think it's because I use autocomplete.

**Comment by ghiscoding** [comment]
I'm reopening this issue since it is now in active development and will completely remove jQueryUI (or at least make it optional) from SlickGrid, it will be replaced by [SortableJS](https://github.com/SortableJS/Sortable) which is much smaller and has much better and smoother UX (it also works nicely with touch).

**Comment by ghiscoding** [comment]
closed by recent PR #695, take a look at the new [3.0.0](https://github.com/6pac/SlickGrid/releases/tag/3.0.0) release
